sesemi/models/heads/loss.py

```diff
@@ -110,7 +118,49 @@ def forward(
         **kwargs,
     ) -> Tensor:
         inputs_u, targets_u = data[self.input_data]
-        x_unlabeled = backbones[self.input_backbone](inputs_u)
-        output_unlabeled = self.fc_unlabeled(x_unlabeled)
-        loss_u = F.cross_entropy(output_unlabeled, targets_u, reduction="none")
+        x_u = backbones[self.input_backbone](inputs_u)
+        output_u = self.fc_unsupervised(x_u)
+        loss_u = F.cross_entropy(output_u, targets_u, reduction="none")
+        return loss_u
+
+
+class EntropyMinimizationLossHead(LossHead):
+    """The entropy minimization loss head.
+    https://papers.nips.cc/paper/2004/file/96f2b50b5d3613adf9c27049b2a888c7-Paper.pdf
+    """
+
+    def __init__(
+        self,
+        input_data: str,
+        input_backbone: str = "backbone",
+        logger: Optional[LightningLoggerBase] = None,
+    ):
+        """Initializes the loss head.
+
+        Args:
+            input_data: The key used to get the unlabeled input data.
+            input_backbone: The key used to get the backbone for feature extraction.
+            logger: An optional PyTorch Lightning logger.
+        """
+        super().__init__(logger)
+        self.input_data = input_data
+        self.input_backbone = input_backbone
+
+    def build(self, backbones: Dict[str, Backbone], heads: Dict[str, Head], **kwargs):
+        self.fc_supervised = heads["supervised_head"]
```

Comment:
If shared heads are added as attributes they will be duplicated in the checkpoint on save. It is best to store only a single copy of the modules (either in the PyTorch `ModuleDict` or as an attribute to the `nn.Module`).